File: backend/src/models/snippet.js
import mongoose from 'mongoose';

const SnippetSchema = new mongoose.Schema({
  userId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  clientId: { type: String, required: true },
  trigger: { type: String, required: true, maxlength: 120 },
  expansion: { type: String, required: true },
  createdAt: { type: Date, required: true },
  updatedAt: { type: Date, required: true },
  deletedAt: { type: Date, default: null },
});

SnippetSchema.index({ userId: 1, clientId: 1 }, { unique: true });
SnippetSchema.index({ userId: 1, updatedAt: 1 });

export const Snippet = mongoose.model('Snippet', SnippetSchema);

File: ios/Privoice/PrivoiceCore/Sources/PrivoiceCore/Models/Snippet.swift
import Foundation
import GRDB

/// A text expansion: when the user dictates `trigger`, the keyboard inserts `expansion`.
/// Stored locally in SQLite and mirrored via `/sync`.
public struct Snippet: Codable, Sendable, Equatable, Hashable, Identifiable, FetchableRecord, PersistableRecord {
    public var clientId: String
    public var trigger: String
    public var expansion: String
    public var createdAt: Date
    public var updatedAt: Date
    public var deletedAt: Date?
    public var syncedAt: Date?

    public var id: String { clientId }

    public init(
        clientId: String = UUID().uuidString.lowercased(),
        trigger: String,
        expansion: String,
        createdAt: Date = Date(),
        updatedAt: Date = Date(),
        deletedAt: Date? = nil,
        syncedAt: Date? = nil
    ) {
        self.clientId = clientId
        self.trigger = trigger
        self.expansion = expansion
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.deletedAt = deletedAt
        self.syncedAt = syncedAt
    }

    public static let databaseTableName = "snippets"

    public enum Columns: String, ColumnExpression {
        case clientId, trigger, expansion, createdAt, updatedAt, deletedAt, syncedAt
    }
}
